src/utils/componentSpec.ts

```diff
@@ -177,6 +177,178 @@ export interface ComponentReference {
   text?: string;
   favorited?: boolean;
 }
+
+export type UnknownComponentReference = ComponentReference | null | undefined;
+
+export type HydratedComponentReference = Omit<
+  ComponentReference,
+  "spec" | "text" | "digest" | "name"
+> & {
+  digest: string;
+  name: string;
+  spec: ComponentSpec;
+  text: string;
+};
+
+type NotMaterializedComponentReference = Omit<
+  ComponentReference,
+  "spec" | "text"
+> & {
+  spec: never;
+  text: never;
+};
+
+export function isNotMaterializedComponentReference(
+  componentReference: UnknownComponentReference,
+): componentReference is NotMaterializedComponentReference {
+  return Boolean(
+    componentReference &&
+      typeof componentReference === "object" &&
+      !componentReference.spec &&
+      !componentReference.text,
+  );
+}
+
+export type DiscoverableComponentReference = Omit<
+  ComponentReference,
+  "digest"
+> & {
+  digest: string;
+};
+
+export function isDiscoverableComponentReference(
+  componentReference: UnknownComponentReference,
+): componentReference is DiscoverableComponentReference {
+  return Boolean(
+    componentReference &&
+      typeof componentReference === "object" &&
+      componentReference.digest !== undefined &&
+      componentReference.digest.length > 0,
+  );
+}
+
+export type LoadableComponentReference = Omit<ComponentReference, "url"> & {
+  url: string;
+};
+
+export function isLoadableComponentReference(
+  componentReference: UnknownComponentReference,
+): componentReference is LoadableComponentReference {
+  return Boolean(
+    componentReference &&
+      typeof componentReference === "object" &&
+      componentReference.url !== undefined &&
+      componentReference.url.length > 0 &&
+      // simple URL validation
+      /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/i.test(componentReference.url),
```

Comment:
given we have url validation in place we can potentially simplify by one line to


```suggestion
      !!componentReference.url &&
      // simple URL validation
      /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/i.test(componentReference.url),
```